### ISSUE TITLE
NullPointerException in Ruby.freezeAndDedupString

### ISSUE THREAD

**Issue by henrikno** [jruby/jruby#8359]
**Environment Information**
  
* JRuby version (`jruby -v`) and command line (flags, JRUBY_OPTS, etc)
`jruby 9.4.7.0 (3.1.4) 2024-04-29 597ff08ac1 OpenJDK 64-Bit Server VM 17.0.11+9 on 17.0.11+9 +indy +jit [x86_64-linux]`
 * Operating system and platform (e.g. `uname -a`)
Running under docker image of logstash 8.14.1.

**Expected Behavior**

No NullPointerException

**Actual Behavior**

Found this stacktrace in the Logstash logs:
```
    "error": "Cannot invoke \"java.lang.ref.WeakReference.get()\" because \"dedupedRef\" is null",
    "exception": "Java::JavaLang::NullPointerException",
    "backtrace": [
      "org.jruby.Ruby.freezeAndDedupString(Ruby.java:4901)",
      "org.jruby.RubyHash.hashKeyString(RubyHash.java:1226)",
      "org.jruby.RubyHash.op_asetForString(RubyHash.java:1204)",
      "org.jruby.RubyHash.fastASetCheckString(RubyHash.java:1155)",
      "org.jruby.RubyHash.op_aset(RubyHash.java:1192)",
      "org.jruby.RubyHash$INVOKER$i$2$0$op_aset.call(RubyHash$INVOKER$i$2$0$op_aset.gen)",
      "org.jruby.runtime.callsite.RefinedCachingCallSite.call(RefinedCachingCallSite.java:108)",
      "usr.share.logstash.vendor.bundle.jruby.$3_dot_1_dot_0.gems.logstash_minus_filter_minus_fingerprint_minus_3_dot_4_dot_4.lib.logstash.filters.fingerprint.invokeOther2:\\=\\{\\}=(/usr/share/logstash/vendor/bundle/jruby/3.1.0/gems/logstash-filter-fingerprint-3.4.4/lib/logstash/filters/fingerprint.rb:194)",
```

Looking at the code it looks like there might be a missing null check in case the wrapper was GCed.
https://github.com/jruby/jruby/blob/4765ec9da51dead3b3a9ad8784c312ae19563c24/core/src/main/java/org/jruby/Ruby.java#L4898-L4901

**Comment by headius** [comment]
You are correct! The reference returned might be GCed between the time we aquire it and the time we get its referenced value. I'll look into a fix.

**Comment by headius** [comment]
I'm a bit confused about this issue.

The result of `computeIfPresent` must be null for this error to happen. That can only happen if there's no existing mapping in the `dedupMap`, but we would not be in this loop unless we had already seen an entry (from `computeIfAbsent` above).

So it seems like the entry gets removed between `computeIfAbsent` and `computeIfPresent`. I'm not sure how that can happen.

I think the following should handle that case, and always unconditionally recomputes the entry whether it is present or not. If it is present and not vacated, it will remain the same reference. If it is not present, the reference we prepared above will be inserted. And each loop we confirm it has not been vacated before returning a result.

```diff
diff --git a/core/src/main/java/org/jruby/Ruby.java b/core/src/main/java/org/jruby/Ruby.java
index 3762f595c8..8a7c76753d 100644
--- a/core/src/main/java/org/jruby/Ruby.java
+++ b/core/src/main/java/org/jruby/Ruby.java
@@ -4895,7 +4895,7 @@ public final class Ruby implements Constantizable {
             // ref is there but vacated, try to replace it until we have a result
             while (true) {
                 wrapper.string = string;
-                dedupedRef = dedupMap.computeIfPresent(wrapper, (key, old) -> old.get() == null ? weakref : old);
+                dedupedRef = dedupMap.compute(wrapper, (key, old) -> old.get() == null ? weakref : old);
 
                 // return result if not vacated
                 unduped = dedupedRef.get();
```

@henrikno I am wondering: perhaps this is happening because a thread is attempting to dedup a string at the same time we are shutting down? Maybe I'm missing something but I don't think it should be possible to have NPE at this point, since we wouldn't be in that loop unless there's an existing entry in `dedupMap`.

@enebo Maybe your eyes can see something mine missed in this logic?

**Comment by henrikno** [comment]
I haven't fully understood the ConcurrentWeakHashMap implementation, but if you call put(x), then GC happens, then get(x), it looks like it could return null, because it calls `e.keyRef.get()`.
https://github.com/jruby/jruby/blob/4765ec9da51dead3b3a9ad8784c312ae19563c24/core/src/main/java/org/jruby/util/collections/ConcurrentWeakHashMap.java#L414C53-L414C67
Note that there's many threads running this code thousands of times, and it was likely during a time with a lot of GC.

**Comment by headius** [comment]
Aha, of course... it gets culled internally by the time we get back to that point so it disappears between the two accesses.

So my patch may work with modification to check for a null entry. Are you able to produce any reproduction for this? (I am assuming the answer is no, because it's going to be a tricky thing to trigger!)

**Comment by headius** [comment]
@henrikno If you are able to test a build of JRuby, I've pushed #8363 to fix this. I think it should be impossible now to get a null reference at the error point.

**Comment by headius** [comment]
@henrikno Merged for release in 9.4.9.0!

**Comment by henrikno** [comment]
Thank you so much for your swift response. I tried to make a reproduction case by making a lot of threads running freezeAndDedupString and gradually allocating more memory until hitting GC issues, but I just hit normal out of memory errors. From reviewing it it should be able to return null though so that should be solved.
